specs/crux/beacon-chain.md

```diff
@@ -0,0 +1,247 @@
+# Crux -- The Beacon Chain
+
+**Notice**: This document is a work-in-progress for researchers and implementers.
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Constants](#constants)
+  - [Domain types](#domain-types)
+- [Preset](#preset)
+  - [Max operations per block](#max-operations-per-block)
+  - [Time parameters](#time-parameters)
+- [Configuration](#configuration)
+  - [Gwei values](#gwei-values)
+- [Containers](#containers)
+  - [Extended containers](#extended-containers)
+    - [`Validator`](#validator)
+    - [`BeaconBlockBody`](#beaconblockbody)
+  - [New containers](#new-containers)
+    - [`DelegationMessage`](#delegationmessage)
+    - [`Delegation`](#delegation)
+- [Beacon chain state transition function](#beacon-chain-state-transition-function)
+  - [Block processing](#block-processing)
+    - [Operations](#operations)
+    - [Delegation](#delegation)
+      - [`get_validator_from_delegation`](#get_validator_from_delegation)
+      - [`process_delegation`](#process_delegation)
+  - [Epoch processing](#epoch-processing)
+    - [`process_delegation_transfers`](#process_delegation_transfers)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This upgrade adds validator delegation. A validator may choose to delegate all its earnings above `MAX_EFFECTIVE_BALANCE` to an existing or a new validator.
+
+## Constants
+
+### Domain types
+
+| Name  | Value |
+| - | - |
+| `DOMAIN_DELEGATION` | `DomainType('0x0A000000')` |
+
+## Preset
+
+### Max operations per block
+
+| Name | Value |
+| - | - |
+| `MAX_DELEGATIONS` | `2**4` (=16) |
+
+### Time parameters
+| Name | Value | Unit | Duration | 
+| - | - | :-: | :-: |
+| `MAX_DELEGATION_INCLUSION_DELAY` | `uint64(2**7)` (=128) | slots | 25.6 minutes | 
+
+## Configuration
+
+### Gwei values
+
+| Name | Value |
+| - | - |
+| `DELEGATION_TRANSACTION_COST` | `Gwei(10**6)` (= 1,000,000) TBD |
+
+
+## Containers
+
+### Extended containers
+
+#### `Validator`
+The `Validator` container gains a new field `delegate` containing the validator index of the delegate validator. This field is initialized at the fork with the same validator index of the current validator. (N.B. implementers may chose to keep a separate map instead)
+
+```python
+class Validator(Container):
+    pubkey: BLSPubkey
+    delegate: ValidatorIndex  # [New in Crux]
```

Comment:
Please insert new fields at the bottom. Otherwise you may break optimizations that assume data ordering either merkelized or serialized